### PR TITLE
Move to CUDA_CALL for nvJPEG, nvJPEG2k, and NPP

### PR DESCRIPTION
**Pull request by JanuszL**
- reworks return status check mechanism to use unified CUDA_CALL
  convention
- adds CUDA_CALL_EX accepting extra error information to be aligned
  with nvJPEG and nvJPEG2k error mechanism

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [x] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- reworks return status check mechanism to use unified CUDA_CALL
  convention
- adds CUDA_CALL_EX accepting extra error information to be aligned
  with nvJPEG and nvJPEG2k error mechanism

#### Additional information
- Affected modules and functionalities:
  - dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h
  - dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h
  - dali/operators/decoder/nvjpeg/nvjpeg_helper.cc
  - dali/operators/decoder/nvjpeg/nvjpeg_helper.h
  - dali/util/npp.h
  - include/dali/core/cuda_error.h

- Key points relevant for the review:
  - if CUDA_CALL_EX makes sense

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A
DALI-2469

### CODE REVIEW

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h [hunk 1]
```diff
@@ -23,29 +23,85 @@
...
+      case NVJPEG2K_STATUS_IMPLEMENTATION_NOT_SUPPORTED:
+        return "API is not supported by the backend.";
+      default:
+        return "< unknown error >";
+    }
+  }
+
+  static std::string Message(nvjpeg2kStatus_t result, const char *details) {
+    if (details && *details) {
+      return make_string("nvjpeg2k error: ", result, " ", ErrorString(result),
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h:71]:
```suggestion
      return make_string("nvJPEG2000 error ", result, ": ", ErrorString(result),
```
The official name of the library is *nvJPEG2000*.
Also, I think that this looks better:
`nvJPEG2000 error 2: Wrong parameter was passed.`
than that:
`nvJPEG2000 error, 2, Wrong parameter was passed.`

Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h]: Done

**Comment by JanuszL** [comment]
!build

File: include/dali/core/cuda_error.h [hunk 2]
```diff
@@ -156,6 +162,11 @@ inline void CUDA_CALL(T status) {
   return dali::cudaResultCheck(status);
 }
 
+template <typename T>
+inline void CUDA_CALL_EX(T status, std::string extra) {
```
Comment (mzient) [inline include/dali/core/cuda_error.h:166]: It's a bad idea to make multiple copies of a `string` for the sake of an unlikely event (an error).
```suggestion
/**
 * @param status A status code that indicates an error or success
 * @param extra Additional information about an error, if it occurs. The
 *              caller should take care not to construct a string solely for the
 *              purpose of this parameter due to performance overhead.
 */
inline void CUDA_CALL_EX(T status, const std::string &extra) {
```

File: include/dali/core/cuda_error.h [hunk 3]
```diff
@@ -98,6 +98,12 @@ inline void cudaResultCheck(Code code) {
     "cudaResultCheck not implemented for this type of status code");
 }
 
+template <typename Code>
+inline void cudaResultCheck(Code code, string extra) {
```
Comment (mzient) [inline include/dali/core/cuda_error.h:102]:
```suggestion
inline void cudaResultCheck(Code code, const std::string &extra) {
```

File: dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h [hunk 4]
```diff
@@ -23,29 +23,85 @@
...
+  switch (status) {
+  case NVJPEG2K_STATUS_SUCCESS:
+    return;
+  default:
+    throw dali::Nvjpeg2kError(status);
+  }
+}
+
+template <>
+inline void cudaResultCheck<nvjpeg2kStatus_t>(nvjpeg2kStatus_t status, string extra) {
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h:96]:
```suggestion
inline void cudaResultCheck<nvjpeg2kStatus_t>(nvjpeg2kStatus_t status, const string &extra) {
```

File: dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h [hunk 5]
```diff
@@ -23,29 +23,85 @@
...
+  switch (status) {
+  case NVJPEG2K_STATUS_SUCCESS:
+    return;
+  default:
+    string extra_info = extra;
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h:101]: What is this for?

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h [hunk 6]
```diff
@@ -19,40 +19,93 @@
...
+  }
+}
+
+template <>
+inline void cudaResultCheck<nvjpegStatus_t>(nvjpegStatus_t status, string extra) {
+  switch (status) {
+  case NVJPEG_STATUS_SUCCESS:
+    return;
+  default:
+    string extra_info = extra;
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg_helper.h:105]: This isn't necessary.

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h [hunk 7]
```diff
@@ -19,40 +19,93 @@
...
+  switch (status) {
+  case NVJPEG_STATUS_SUCCESS:
+    return;
+  default:
+    throw dali::NvjpegError(status);
+  }
+}
+
+template <>
+inline void cudaResultCheck<nvjpegStatus_t>(nvjpegStatus_t status, string extra) {
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg_helper.h:100]:
```suggestion
inline void cudaResultCheck<nvjpegStatus_t>(nvjpegStatus_t status, const std::string &extra) {
```

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h [hunk 8]
```diff
@@ -19,40 +19,93 @@
...
+  switch (status) {
+  case NVJPEG_STATUS_SUCCESS:
+    return;
+  default:
+    string extra_info = extra;
+    throw dali::NvjpegError(status, extra_info.c_str());
```
Comment (mzient) [inline dali/operators/decoder/nvjpeg/nvjpeg_helper.h:106]:
```suggestion
    throw dali::NvjpegError(status, extra.c_str());
```

File: include/dali/core/cuda_error.h (hunk 3 above)
Comment (JanuszL) [reply include/dali/core/cuda_error.h]: Done

File: dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h (hunk 4 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h]: Done

File: dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h (hunk 5 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h]: Removed

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h (hunk 6 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_helper.h]: Removed

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h (hunk 7 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_helper.h]: Done

File: dali/operators/decoder/nvjpeg/nvjpeg_helper.h (hunk 8 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/nvjpeg_helper.h]: Done

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Pull request merged**